File: 680.valid-palindrome-ii.java
/*
 * @lc app=leetcode id=680 lang=java
 *
 * [680] Valid Palindrome II
 */

// @lc code=start
class Solution {
    boolean deleteFlag = false;

    public boolean validPalindrome(String s) {

        int left = 0, right = s.length() - 1;
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                if (deleteFlag == false) {
                    String deleteLeft = removeCharAt(s, left);
                    String deleteRight = removeCharAt(s, right);
                    deleteFlag = true;
                    return validPalindrome(deleteLeft) || validPalindrome(deleteRight);
                } else {
                    return false;
                }
            } else {
                left++;
                right--;
            }
        }
        return true;
    }

    public static String removeCharAt(String s, int pos) {
        return s.substring(0, pos) + s.substring(pos + 1);
    }
}
// @lc code=end

File: 680.valid-palindrome-ii.py
#
# @lc app=leetcode id=680 lang=python3
#
# [680] Valid Palindrome II
#

# @lc code=start

class Solution:
    def validPalindrome(self, s: str) -> bool:
        def isValid(chars):
            n=len(chars)
            left,right=0,n-1
            while left<right:
                if chars[left]!=chars[right]:
                    return False
                left+=1
                right-=1

            return True

        n=len(s)
        left,right=0,n-1
        while left<right:
            if s[left]!=s[right]:
                isValidDelLeft=isValid(s[left+1:right+1])
                isValidDelRight=isValid(s[left:right])
                return isValidDelLeft or isValidDelRight
            left+=1
            right-=1

        return True
        
# @lc code=end

